### PR TITLE
cpu/native: fix race in thread_yield_higher()

### LINKED ISSUE #6123: gnrc: crash with (excessive) traffic in native
#2071 seems to have fixed native interrupt handling. But I can still crash a riot native instance by flood pinging.

Steps to reproduce:

1. run gnrc_networking
2. start a couple (5-10) ```ping -f <riot-ip>%tap0```
3. wait some minutes

Sometimes it takes 30sec, sometimes 5min, but riot reliably crashes.

The problem seems somewhere in gnrc, as my other stack runs fine since #2071 is merged.
Here's a backtrace:

```
Program received signal SIGSEGV, Segmentation fault.
0x0804df49 in gnrc_pktbuf_start_write ()
(gdb) bt
#0  0x0804df49 in gnrc_pktbuf_start_write ()
#1  0x0805c0ae in _send ()
#2  0x0805ba21 in _event_loop ()
#3  0xf7dba58b in makecontext () from /usr/lib32/libc.so.6
#4  0x00000000 in ?? ()
(gdb)
```

*Edit:* summary of reasons why this might not be a GNRC issue in https://github.com/RIOT-OS/RIOT/issues/6123#issuecomment-457848344.

### LINKED ISSUE #10881: core: `msg_receive()` on native sometimes returns without `msg` being re-set
#### Description
Coming from #6123 I was able to track down the issue to the fact that `msg_receive()` in some corner-cases seems to return without the out parameter being rewritten. I wasn't able to pin-point the exact issue yet and I'm unsure if it is only an issue on `native`.

#### Steps to reproduce the issue
I applied the following patch to current master (bdd2d52fd57c8cc87f9705b2e31c32f58a123a06, might not apply to later versions, please then try to change manually):

```diff
diff --git a/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c b/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c
index feb2e8f10..0c3fa97c4 100644
--- a/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c
+++ b/sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c
@@ -180,6 +180,7 @@ static void *_event_loop(void *args)
     /* start event loop */
     while (1) {
         DEBUG("ipv6: waiting for incoming message.\n");
+        memset(&msg, 0, sizeof(msg));
         msg_receive(&msg);
 
         switch (msg.type) {
@@ -220,6 +221,7 @@ static void *_event_loop(void *args)
                 gnrc_ipv6_nib_handle_timer_event(msg.content.ptr, msg.type);
                 break;
             default:
+                printf("ipv6: unknown message type 0x%04x\n", msg.type);
                 break;
         }
     }
```

and ran `gnrc_networking`. I then tried to ping the node as described in #10875:

```sh
sudo true; for _ in $(seq 10); do sudo ping -s0 -f "fe80::...%tapbr0" & done
```

*Edit:* Alternatively just run the tests in #10908.

#### Expected results
```
ipv6: unknown message type 0x0000
```

should never show up.

#### Actual results
```
ipv6: unknown message type 0x0000
```

shows up with some regularity

#### Versions
bdd2d52fd57c8cc87f9705b2e31c32f58a123a06 on a somewhat recent Arch as of writing this issue.

```
Operating System Environment
-----------------------------
       Operating System: "Arch Linux" 
                 Kernel: Linux 4.20.3-arch1-1-ARCH x86_64 unknown

Installed compiler toolchains
-----------------------------
             native gcc: gcc (GCC) 8.2.1 20181127
      arm-none-eabi-gcc: arm-none-eabi-gcc (Arch Repository) 8.2.0
                avr-gcc: avr-gcc (GCC) 8.2.0
       mips-mti-elf-gcc: missing
             msp430-gcc: missing
   riscv-none-embed-gcc: missing
   xtensa-esp32-elf-gcc: missing
   xtensa-lx106-elf-gcc: missing
                  clang: clang version 7.0.1 (tags/RELEASE_701/final)

Installed compiler libs
-----------------------
   arm-none-eabi-newlib: "3.0.0"
    mips-mti-elf-newlib: missing
riscv-none-embed-newlib: missing
xtensa-esp32-elf-newlib: missing
xtensa-lx106-elf-newlib: missing
               avr-libc: "2.0.0" ("20150208")

Installed development tools
---------------------------
                  cmake: cmake version 3.13.3
               cppcheck: missing
                doxygen: 1.8.15
                 flake8: 3.6.0 (mccabe: 0.6.1, pycodestyle: 2.4.0, pyflakes: 2.0.0) CPython 3.7.2 on Linux
                    git: git version 2.20.1
                   make: GNU Make 4.2.1
                openocd: missing
                 python: Python 3.7.2
                python2: Python 2.7.15
                python3: Python 3.7.2
             coccinelle: missing
```

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

    Error case:
    1. thread_yield_higher() stores the thread's ucontext
    2. creates an "isr ucontext" for isr_thread_yield, switches to it
    
    Case 1: no signals are pending, continues in isr_thread_yield()
    3a. sched_run is called
    4a. return to sched_active_thread ucontext
    
    Case 2: signals pending (the crashing scenario), continues in native_irq_handler()
    3b. handles signals
    4b. if sched_context_switch_request is set, call sched_run
    5b. return to sched_active_thread ucontext
    
    4b misses the call to sched_run(), leading to a possible return into a
    non-ready thread.

Fixes #10881.

### Testing procedure

See #10881 and #10908

### Issues/PRs references

Fixes #10881.
Fixes #6123.

### CODE REVIEW

**Comment by kaspar030** [comment]
This is pretty central in native, s I guess a couple of eyes need to take a look...

**Comment by kaspar030** [comment]
@LudwigKnuepfer maybe you could take a look?

**Comment by kaspar030** [comment]
This fixes #6123 for me!

**Comment by miri64** [comment]
I'm rather wondering if

```diff
diff --git a/cpu/native/native_cpu.c b/cpu/native/native_cpu.c
index 2629e55..79534e3 100644
--- a/cpu/native/native_cpu.c
+++ b/cpu/native/native_cpu.c
@@ -142,7 +142,7 @@ void isr_cpu_switch_context_exit(void)
     ucontext_t *ctx;
 
     DEBUG("isr_cpu_switch_context_exit\n");
-    if ((sched_context_switch_request == 1) || (sched_active_thread == NULL)) {
+    if (sched_active_thread == NULL) {
         sched_run();
     }
```

wouldn't be a more valid fix. Because this is basically also creating the situation you are creating:

- `isr_cpu_switch_context_exit()` is only called by `cpu_switch_context_exit()` (either with a context switch to ISR when not in ISR, or directly when in ISR)
- `cpu_switch_context_exit()` is called when all signal handlers were run in `native_irq_handler()` and in `sched_task_exit`, which is also called when you end a thread (so the user probably also wants to switch contexts), and also `kernel_init` for some reason...
- so `isr_switch_context_exit()` should only be called when a context switch was requested, right?

**Comment by miri64** [comment]
Some historic research:

- The check in `isr_cpu_switch_context_exit()` was introduced by @LudwigKnuepfer in https://github.com/RIOT-OS/RIOT/pull/229
- The setting of of `sched_context_switch_request` in the else branch was introduced by @kaspar030 in https://github.com/RIOT-OS/RIOT/pull/6660

**Comment by kaspar030** [comment]
> ```if (sched_active_thread == NULL)```
> wouldn't be a more valid fix.

I don't think so, as it would not call ```sched_run()``` anymore (sched_active_thread is never NULL unless threading is starting). *always* calling sched_run there would also be a "fix" but it would probably call the scheduler more often than needed.

**Comment by miri64** [comment]
The fix I proposed causes the scheduler only when the thread is exited, so this obviously was bullshit. Also, since the title of #229 where this check was introduced mentions segfaults we probably shouldn't mess with that.

**Comment by kaspar030** [comment]
> This fixes #6123 for me!

stable for 30min now, ```sudo true; for _ in $(seq 50); do sudo ping -s1392 -f "${TARGET}" & done```.

This bug has haunted my benchmarking efforts on native for years!

**Comment by miri64** [comment]
Penny for your thoughts on this though https://github.com/RIOT-OS/RIOT/pull/6660#issuecomment-283609379

**Comment by miri64** [comment]
> This bug has haunted my benchmarking efforts on native for years!

Next time use some elbow grease like I did ;D.

**Comment by kaspar030** [comment]
> Penny for your thoughts on this though [#6660 (comment)](https://github.com/RIOT-OS/RIOT/pull/6660#issuecomment-283609379)

You mean ```at the end of native's IRQ handler (native_irq_handler()) the function cpu_switch_context_exit() gets called.```?

I think the naming is a little off. 

```cpu_switch_context_exit()``` basically jumps to another context without saving (or better: with completely ignoring)  the current one. That is also why thread_yield_higher(), after saving the context (via swapcontext()), ends up there, that's why "kernel_init()" uses it to kick off scheduling and why the function is put into a new thread's stack as return value so when a thread function exits, it ends up calling.

**Comment by miri64** [comment]
Ok my final assessment for today for this PR is that `sched_run` might be called unnecessarily in some cases, which on native I can live with. I'll run some tests tomorrow and then give my final verdict.

**Comment by miri64** [comment]
I tested most thread-related tests on `native` locally again. All succeeded. The rest I leave to Murdock.

**Comment by miri64** [comment]
I was however not able find an isolated test case yet :-(.

**Comment by miri64** [comment]
Maybe because I don't fully understand how and when `sched_context_switch_request ` can be 0 in and when it is 1.

**Comment by miri64** [comment]
Maybe @OlegHahm can also have a look, since he reviewed #6660?

**Comment by miri64** [comment]
> Except `native` only `atmega_common` seems to do that and also only in the else branch of an `irq_is_in() == 0` check:
> 
> https://github.com/RIOT-OS/RIOT/blob/782b181b63ce16684279ccac2f84d890727360c4/cpu/atmega_common/thread_arch.c#L219-L230
> 
> (*edit: but maybe this was just mapped after `native`*)

Was introduced in https://github.com/RIOT-OS/RIOT/pull/8904 so summon @ZetaR60 as well.

**Comment by kaspar030** [comment]
> Maybe because I don't fully understand how and when `sched_context_switch_request ` can be 0 in and when it is 1.

TL;DR ```sched_context_switch_request``` is used to request a (necessarily deferred) context switch from within an ISR.

A context switch can be triggered either from the running task or from within an ISR.
In the former case, it can be done right away: save the current context, run ```sched_run()``` to find the next thread, then restore that.

In ISRs, the context switch cannot be done right away. See e.g. this (pseudocode):

```
void foobar_isr() {
  do_stuff();
  if (whatever > FOO) {
    mutex_unlock(mutex_a_thread_is_waiting_on);
  } 
  do_more_stuff(); 
}
```

In ISR context, the ISR must finish first. It is not possible to "pause" ISR context, switch to the thread, then "continue" the ISR context. ISRs must finish. So the thread that would probably be unlocked by the mutec_unlock() call is merely put on the runqueue, but we need to somehow pass along the information that a context switch might be necessary.

This is what "sched_context_switch_request" is used for: pass along the information that a context switch might be necessary after an ISR has finished.

Thing is, different platforms have different concepts.

E.g., on Cortex-M, there's PendSV, which will always be called last. Also, there are the banked ISR registers (a second set of registers), thus for serving an ISR, no manual context saving is necessary. Only when, after all ISRs have run, a different user thread should continue to run, the current context needs to be saved, the next selected and then restored, through fiddling with the user mode stack. ```sched_context_switch_request``` is used as a flag to let the architecture code, at the end of an ISR, know that this work should be done and thus PendSV should be triggered.

Cortex-M checks the flag here:

https://github.com/RIOT-OS/RIOT/blob/782b181b63ce16684279ccac2f84d890727360c4/cpu/cortexm_common/include/cpu.h#L185-L190

Other platforms do it differently, but all are using the flag.

**Comment by miri64** [comment]
Ok, I think most of that I was able to figure out myself, so thanks for the affirmation and putting it in words. However, I still don't really am able to track how the case before was wrong / let to invalid unblocking so that is why I don't feel 100% confident in this bug fix.

**Comment by kaspar030** [comment]
> However, I still don't really am able to track how the case before was wrong / let to invalid unblocking so that is why I don't feel 100% confident in this bug fix.

I'll try to put it in different words.

Usually on thread_yield_higher() on native, when called outside of an ISR, an "isr context" is created with "makecontext" which executes "isr_thread_yield". "swapcontext()" saves the thread context and jumps into the newly created "isr context", launching "isr_thread_yield".

Now, if there are no signals pending, the scheduler is run, then a context switch to the then-current sched_active_thread is initiated. All is good.

But it is possible that signals queue up (after irq_disable in thread_yield_higher()), making "isr_thread_yield()" call "native_irq_handler()". That function does not return, as it returns to thread context by itself, after handling ISRs. The jumping-back to thread context is almost identical to how "isr_thread_yield()" does it (if there'd been no signals pending). The main difference is that "native_irq_handler()" does not always call "sched_run()", but only if "sched_context_switch_request" was set.

This lead to the crash at hand: msg_receive() removes the current thread from the runqueue, then calls thread_yield_higher() *expecting* to be scheduled away. But in the off chance that a signal pops up at the wrong time (and the ISR does not set sched_context_switch_request=1), the scheduler doesn't get called, and msg_receive() returns without receiving.

The fix works by *always* setting "sched_context_switch_request" to 1, so should "isr_thread_yield" degrade to "native_irq_handler", the scheduler is still called.

**Comment by kaspar030** [comment]
>  (trigger a timer in very short intervals [which causes a SIGALRM on native], while a message exchange between two threads is happening), they just were not able to trigger the race condition.

Actually yes, if the timer does not trigger a context switch (e.g., no mutex unlock or msg send)!

**Comment by miri64** [comment]
> Actually yes, if the timer does not trigger a context switch (e.g., no mutex unlock or msg send)!

Aha, that was what missing then :-).

**Comment by miri64** [comment]
Ok, I tried that, but still not really able to reproduce the issue. Probably the receiving thread needs to do a little more than just printing something.

**Comment by miri64** [comment]
https://github.com/miri64/RIOT/blob/tests/new/i10881/tests/thread_msg_block_race/main.c

**Comment by kaspar030** [comment]
> Aha, that was what missing then :-).

Ah, sorry, I didn't get you. No context switch was right...

**Comment by miri64** [comment]
Ok, I did that now and also replaced `msg_try_send()` with `msg_send()` (otherwise the main thread starves). However, now sending stops after 3 times trying... :(

**Comment by miri64** [comment]
I got it! Let me write a quick test script and I provide a PR.

**Comment by emmanuelsearch** [comment]
@kaspar030 so we're green?

**Comment by kaspar030** [comment]
> @kaspar030 so we're green?

Yes!

Many many thanks to @miri64 for reducing #6123 to ```msg_receive()``` and debugging this!

> (which means start preparing a backport hehe ;)

Could someone? Isn't there a script?

**Comment by miri64** [comment]
Yepp, but not merged yet https://github.com/RIOT-OS/RIOT/pull/8968 (you can still use it though from @bergzand's branch)

**Comment by miri64** [comment]
Backport provided in #10921
